python/lsst/ctrl/mpexec/job_report.py

```diff
@@ -0,0 +1,188 @@
+# This file is part of ctrl_mpexec.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import annotations
+
+__all__ = ["ExecutionStatus", "JobReport", "QuantumReport"]
+
+import enum
+import shlex
+import sys
+from typing import Dict, List, Optional
+
+from lsst.daf.butler import DataCoordinate, DataId, DataIdValue
+from lsst.utils.introspection import get_full_type_name
+from pydantic import BaseModel
+
+
+def _cmdLine() -> str:
+    """Return command line for current process."""
+    return shlex.join(sys.argv)
+
+
+def _serializeDataId(dataId: DataId) -> Dict[str, DataIdValue]:
+    if isinstance(dataId, DataCoordinate):
+        return dataId.byName()
+    else:
+        return dataId  # type: ignore
+
+
+class ExecutionStatus(enum.Enum):
+    """Possible values for job execution status.
+
+    Status `FAILURE` is set if one or more tasks failed. Status `TIMEOUT` is
+    set if there are no failures but one or more tasks timed out. Timeouts can
+    only be detected in multi-process mode, child task is killed on timeout
+    and usually should have non-zero exit code.
+    """
+
+    SUCCESS = "success"
+    FAILURE = "failure"
+    TIMEOUT = "timeout"
```

Comment:
When running pipelines_check, I made isr fail due by chmod'ing the raw file.  In the log, I got 

> INFO 2022-03-02T18:41:52.993-06:00 lsst.ctrl.mpexec.mpGraphExecutor ()(mpGraphExecutor.py:451) - Executed 0 quanta successfully, 1 failed and 2 remain out of total 3 quanta.
> 
> ERROR 2022-03-02T18:41:52.993-06:00 lsst.ctrl.mpexec.mpGraphExecutor ()(mpGraphExecutor.py:409) - Upstream job failed for task <TaskDef(lsst.pipe.tasks.characterizeImage.CharacterizeImageTask, label=characterizeImage) dataId={instrument: 'HSC', detector: 10, visit: 903342, ...}>, skipping this task.
> ERROR 2022-03-02T18:41:52.993-06:00 lsst.ctrl.mpexec.mpGraphExecutor ()(mpGraphExecutor.py:409) - Upstream job failed for task <TaskDef(lsst.pipe.tasks.calibrate.CalibrateTask, label=calibrate) dataId={instrument: 'HSC', detector: 10, visit: 903342, ...}>, skipping this task.

And in the jobreport.json file I got entries for all 3 quanta.  The report for isr made sense.  The reports for characterizeImage and calibrate said `"status": "failure",`  Were these downstream ones supposed to be included in the report?  If yes, can we have a different status (e.g., pruned).   Also should we have a different status for a quantum that is skipped because previously executed successfully?

Reply:
I think I decided not to include skipped quanta, but looking at the code now they are included for in-process running (I guess you ran above test with -j1?) In multi-process they should not appear. I have no big preference, if you prefer I can add "pruned" status for those and always include them.

Reply:
I added SKIPPED="skipped" status, setting it for all skipped dependencies.